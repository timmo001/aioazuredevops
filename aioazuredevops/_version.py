"""Version helper for aioazuredevops.

Expose a simple string value for the current package version,
to remove the dependency on incremental and its auto-generated file.
"""

from __future__ import annotations

try:
    # Prefer to derive version from setup.py via importlib.metadata when installed
    from importlib.metadata import version as _dist_version

    __version__ = _dist_version("aioazuredevops")
except Exception:  # pragma: no cover - fallback to static string if not installed
    # Keep in sync with setup.py
    __version__ = "2.2.2.dev0"

__all__ = ["__version__"]
