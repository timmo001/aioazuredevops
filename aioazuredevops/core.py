"""
DevOps Core
https://docs.microsoft.com/en-gb/rest/api/azure/devops/core/?view=azure-devops-rest-6.0
"""
from datetime import datetime


class DevOpsLinks:
    """Links."""

    def __init__(self, linkSelf: str = None, linkCollection: str = None, linkWeb: str = None):
        """Initialize"""
        self.self = linkSelf
        self.collection = linkCollection
        self.web = linkWeb


class DevOpsTeam:
    """DevOps Team."""

    def __init__(self, id: str, name: str, url: str = None):
        """Initialize"""
        self.id = id
        self.name = name
        self.url = url


class DevOpsProject:
    """A DevOps Project"""

    def __init__(
        self,
        id: str,
        name: str,
        description: str = None,
        url: str = None,
        state: str = None,
        revision: int = None,
        visibility: str = None,
        last_updated: datetime = None,
        default_team: DevOpsTeam = None,
        links: DevOpsLinks = None,
    ):
        """Initialize"""
        self.id = id
        self.name = name
        self.description = description
        self.url = url
        self.state = state
        self.revision = revision
        self.visibility = visibility
        self.last_updated = last_updated
        self.default_team = default_team
        self.links = links
