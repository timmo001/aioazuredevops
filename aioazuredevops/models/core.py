"""DevOps Core.

https://docs.microsoft.com/en-gb/rest/api/azure/devops/core/?view=azure-devops-rest-7.2-preview
"""

from dataclasses import dataclass
from datetime import datetime


@dataclass
class ProcessTemplate:
    """Azure DevOps project process template."""

    template_name: str
    template_type_id: str


@dataclass
class VersionControl:
    """Azure DevOps project version control."""

    source_control_type: str
    git_enabled: str
    tfvc_enabled: str


@dataclass
class Capabilities:
    """Azure DevOps project capabilities."""

    process_template: ProcessTemplate
    versioncontrol: VersionControl


@dataclass
class DefaultTeam:
    """Azure DevOps project default team."""

    id: str
    name: str
    url: str


@dataclass
class LinkCollection:
    """Azure DevOps project collection."""

    href: str


@dataclass
class Links:
    """Azure DevOps project links."""

    links_self: LinkCollection
    collection: LinkCollection
    web: LinkCollection


@dataclass
class Project:
    """Azure DevOps project."""

    id: str
    name: str
    description: str
    url: str
    state: str
    revision: int
    visibility: str
    capabilities: Capabilities | None = None
    links: Links | None = None
    default_team: DefaultTeam | None = None
    last_update_time: datetime | None = None
