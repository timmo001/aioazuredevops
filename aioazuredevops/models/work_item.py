"""Work items in a project.

https://docs.microsoft.com/en-gb/rest/api/azure/devops/wit/work-items/list?view=azure-devops-rest-7.2-preview
"""

from dataclasses import dataclass
from datetime import datetime
from typing import Any
from uuid import UUID

from . import ListResult


@dataclass
class WorkItemAvatar:
    """Work item avatar."""

    href: str


@dataclass
class WorkItemLinks:
    """Work item links."""

    avatar: WorkItemAvatar


@dataclass
class WorkItemUser:
    """Work item user."""

    display_name: str
    url: str
    id: UUID
    unique_name: Any
    image_url: Any
    descriptor: str
    links: WorkItemLinks | None = None


@dataclass
class WorkItemFields:
    """Azure DevOps work item fields."""

    area_path: str
    team_project: str
    iteration_path: str
    work_item_type: str
    state: str
    reason: str
    assigned_to: WorkItemUser | None
    created_date: datetime
    created_by: WorkItemUser | None
    changed_date: datetime
    changed_by: WorkItemUser | None
    comment_count: int
    title: str
    microsoft_vsts_common_state_change_date: datetime
    microsoft_vsts_common_priority: int


@dataclass
class WorkItem:
    """Azure DevOps work item."""

    id: int
    rev: int
    fields: WorkItemFields
    url: str


type WorkItemsResult = ListResult[WorkItem]
