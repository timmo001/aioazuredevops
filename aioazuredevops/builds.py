"""
DevOps Builds
https://docs.microsoft.com/en-gb/rest/api/azure/devops/build/builds?view=azure-devops-rest-6.0
"""
from aioazuredevops.core import DevOpsProject


class DevOpsBuildLinks:
    """DevOps build links."""

    def __init__(
        self,
        l_self: str = None,
        web: str = None,
        source_version_display_uri: str = None,
        timeline: str = None,
        badge: str = None,
    ):
        """Initialize"""
        self.self = l_self
        self.web = web
        self.source_version_display_uri = source_version_display_uri
        self.timeline = timeline
        self.badge = badge


class DevOpsBuildDefinition:
    """DevOps build definition."""

    def __init__(
        self,
        id: int,
        name: str,
        url: str = None,
        path: str = None,
        type: str = None,
        queue_status: str = None,
        revision: int = None,
    ):
        self.id = id
        self.name = name
        self.url = url
        self.path = path
        self.type = type
        self.queue_status = queue_status
        self.revision = revision


class DevOpsBuild:
    """DevOps build."""

    def __init__(
        self,
        id: int,
        build_number: str = None,
        status: str = None,
        result: str = None,
        source_branch: str = None,
        source_version: str = None,
        priority: str = None,
        reason: str = None,
        queue_time: str = None,
        start_time: str = None,
        finish_time: str = None,
        definition: DevOpsBuildDefinition = None,
        project: DevOpsProject = None,
        links: DevOpsBuildLinks = None,
    ):
        """Initialize"""
        self.id = id
        self.links = links
        self.build_number = build_number
        self.status = status
        self.result = result
        self.source_branch = source_branch
        self.source_version = source_version
        self.priority = priority
        self.reason = reason
        self.queue_time = queue_time
        self.start_time = start_time
        self.finish_time = finish_time
        self.definition = definition
        self.project = project
