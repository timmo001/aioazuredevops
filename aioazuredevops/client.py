"""Get data from the Azure DevOps API"""
from datetime import datetime

import aiohttp

from aioazuredevops.builds import DevOpsBuild, DevOpsBuildDefinition, DevOpsBuildLinks
from aioazuredevops.core import DevOpsLinks, DevOpsProject, DevOpsTeam
from aioazuredevops.wiql import DevOpsWiqlColumn, DevOpsWiqlResult, DevOpsWiqlWorkItem
from aioazuredevops.work_item import (
    DevOpsWorkItem,
    DevOpsWorkItemAvatar,
    DevOpsWorkItemLinks,
    DevOpsWorkItemUser,
    DevOpsWorkItemValue,
    DevOpsWorkItemValueFields,
)


class DevOpsClient:
    """Client for Azure DevOps."""

    def __init__(self) -> None:
        """Initilalize."""
        self._authorized = False
        self._pat = None

    @property
    def authorized(self):
        """Is the client authorized."""
        return self._authorized

    @property
    def pat(self):
        """Get the PAT."""
        return self._pat

    async def get(
        self,
        session: aiohttp.ClientSession,
        url: str,
    ) -> aiohttp.ClientResponse:
        """Runs a GET request and returns response"""
        if self._pat is None:
            return await session.get(url)
        return await session.get(
            url,
            headers={"Authorization": aiohttp.BasicAuth("", self._pat).encode()},
        )

    async def post(
        self,
        session: aiohttp.ClientSession,
        url: str,
        data: dict,
    ) -> aiohttp.ClientResponse:
        """Runs a POST request and returns response"""
        if self._pat is None:
            return await session.post(
                url,
                json=data,
            )
        return await session.post(
            url,
            json=data,
            headers={"Authorization": aiohttp.BasicAuth("", self._pat).encode()},
        )

    async def authorize(
        self,
        pat: str,
        organization: str,
    ) -> None:
        """Authenticate."""
        async with aiohttp.ClientSession() as session:
            self._pat = pat
            response: aiohttp.ClientResponse = await self.get(
                session, f"https://dev.azure.com/{organization}/_apis/projects"
            )
            if response.status == 200:
                self._authorized = True
            else:
                self._authorized = False

    async def get_project(
        self,
        organization: str,
        project: str,
    ) -> DevOpsProject:
        """Get Azure DevOps project."""
        async with aiohttp.ClientSession() as session:
            response: aiohttp.ClientResponse = await self.get(
                session,
                f"https://dev.azure.com/{organization}/_apis/projects/{project}",
            )
            if response.status != 200:
                return None
            json = await response.json()
            if json is None:
                return None

            return DevOpsProject(
                json["id"],
                json["name"],
                json["description"] if "description" in json else None,
                json["url"] if "url" in json else None,
                json["state"] if "state" in json else None,
                json["revision"] if "revision" in json else None,
                json["visibility"] if "visibility" in json else None,
                datetime.strptime(json["lastUpdateTime"], "%Y-%m-%dT%H:%M:%S.%fZ")
                if "lastUpdateTime" in json
                else None,
                DevOpsTeam(
                    json["defaultTeam"]["id"],
                    json["defaultTeam"]["name"],
                    json["defaultTeam"]["url"],
                )
                if "defaultTeam" in json
                else None,
                DevOpsLinks(
                    json["_links"]["self"]["href"],
                    json["_links"]["collection"]["href"],
                    json["_links"]["web"]["href"],
                )
                if "_links" in json
                else None,
            )

    async def get_builds(
        self,
        organization: str,
        project: str,
        parameters: str,
    ) -> list[DevOpsBuild]:
        """Get Azure DevOps builds."""
        async with aiohttp.ClientSession() as session:
            response: aiohttp.ClientResponse = await self.get(
                session,
                f"https://dev.azure.com/{organization}/{project}/_apis/build/builds{parameters}",
            )
            if response.status != 200:
                return None
            json = await response.json()
            if json is None:
                return None

            builds = []
            for build in json["value"]:
                builds.append(
                    DevOpsBuild(
                        build["id"],
                        build["buildNumber"] if "buildNumber" in build else None,
                        build["status"] if "status" in build else None,
                        build["result"] if "result" in build else None,
                        build["sourceBranch"] if "sourceBranch" in build else None,
                        build["sourceVersion"] if "sourceVersion" in build else None,
                        build["priority"] if "priority" in build else None,
                        build["reason"] if "reason" in build else None,
                        build["queueTime"] if "queueTime" in build else None,
                        build["startTime"] if "startTime" in build else None,
                        build["startTime"] if "startTime" in build else None,
                        DevOpsBuildDefinition(
                            build["definition"]["id"],
                            build["definition"]["name"],
                            build["definition"]["url"] if "url" in build["definition"] else None,
                            build["definition"]["path"] if "path" in build["definition"] else None,
                            build["definition"]["type"] if "type" in build["definition"] else None,
                            build["definition"]["queueStatus"]
                            if "queueStatus" in build["definition"]
                            else None,
                            build["definition"]["revision"]
                            if "revision" in build["definition"]
                            else None,
                        )
                        if "definition" in build
                        else None,
                        DevOpsProject(
                            build["project"]["id"],
                            build["project"]["name"],
                            build["project"]["description"]
                            if "description" in build["project"]
                            else None,
                            build["project"]["url"] if "url" in build["project"] else None,
                            build["project"]["state"] if "state" in build["project"] else None,
                            build["project"]["revision"]
                            if "revision" in build["project"]
                            else None,
                            build["project"]["visibility"]
                            if "visibility" in build["project"]
                            else None,
                            datetime.strptime(
                                build["project"]["lastUpdateTime"],
                                "%Y-%m-%dT%H:%M:%S.%fZ",
                            )
                            if "lastUpdateTime" in build["project"]
                            else None,
                        )
                        if "project" in build
                        else None,
                        DevOpsBuildLinks(
                            build["_links"]["self"]["href"] if "self" in build["_links"] else None,
                            build["_links"]["web"]["href"] if "web" in build["_links"] else None,
                            build["_links"]["sourceVersionDisplayUri"]["href"]
                            if "sourceVersionDisplayUri" in build["_links"]
                            else None,
                            build["_links"]["timeline"]["href"]
                            if "timeline" in build["_links"]
                            else None,
                            build["_links"]["badge"]["href"]
                            if "badge" in build["_links"]
                            else None,
                        )
                        if "_links" in build
                        else None,
                    )
                )

            return builds

    async def get_build(
        self,
        organization: str,
        project: str,
        build_id: int,
    ) -> DevOpsBuild:
        """Get Azure DevOps builds."""
        async with aiohttp.ClientSession() as session:
            response: aiohttp.ClientResponse = await self.get(
                session,
                f"https://dev.azure.com/{organization}/{project}/_apis/build/builds/{build_id}",
            )
            if response.status != 200:
                return None
            build = await response.json()
            if build is None:
                return None

            return DevOpsBuild(
                build["id"],
                build["buildNumber"] if "buildNumber" in build else None,
                build["status"] if "status" in build else None,
                build["result"] if "result" in build else None,
                build["sourceBranch"] if "sourceBranch" in build else None,
                build["sourceVersion"] if "sourceVersion" in build else None,
                build["priority"] if "priority" in build else None,
                build["reason"] if "reason" in build else None,
                build["queueTime"] if "queueTime" in build else None,
                build["startTime"] if "startTime" in build else None,
                build["startTime"] if "startTime" in build else None,
                DevOpsBuildDefinition(
                    build["definition"]["id"],
                    build["definition"]["name"],
                    build["definition"]["url"] if "url" in build["definition"] else None,
                    build["definition"]["path"] if "path" in build["definition"] else None,
                    build["definition"]["type"] if "type" in build["definition"] else None,
                    build["definition"]["queueStatus"]
                    if "queueStatus" in build["definition"]
                    else None,
                    build["definition"]["revision"] if "revision" in build["definition"] else None,
                )
                if "definition" in build
                else None,
                DevOpsProject(
                    build["project"]["id"],
                    build["project"]["name"],
                    build["project"]["description"] if "description" in build["project"] else None,
                    build["project"]["url"] if "url" in build["project"] else None,
                    build["project"]["state"] if "state" in build["project"] else None,
                    build["project"]["revision"] if "revision" in build["project"] else None,
                    build["project"]["visibility"] if "visibility" in build["project"] else None,
                    datetime.strptime(
                        build["project"]["lastUpdateTime"],
                        "%Y-%m-%dT%H:%M:%S.%fZ",
                    )
                    if "lastUpdateTime" in build["project"]
                    else None,
                )
                if "project" in build
                else None,
                DevOpsBuildLinks(
                    build["_links"]["self"]["href"] if "self" in build["_links"] else None,
                    build["_links"]["web"]["href"] if "web" in build["_links"] else None,
                    build["_links"]["sourceVersionDisplayUri"]["href"]
                    if "sourceVersionDisplayUri" in build["_links"]
                    else None,
                    build["_links"]["timeline"]["href"] if "timeline" in build["_links"] else None,
                    build["_links"]["badge"]["href"] if "badge" in build["_links"] else None,
                )
                if "_links" in build
                else None,
            )

    async def get_work_items_ids_all(
        self,
        organization: str,
        project: str,
    ) -> DevOpsWiqlResult:
        """Get Azure DevOps work item ids from wiql."""
        async with aiohttp.ClientSession() as session:
            response: aiohttp.ClientResponse = await self.post(
                session,
                f"https://dev.azure.com/{organization}/{project}/_apis/wit/wiql?api-version=6.0",
                {
                    "query": "Select [System.Id] From WorkItems",
                },
            )
            if response.status != 200:
                return None
            data = await response.json()
            if data is None:
                return None

            return DevOpsWiqlResult(
                query_type=data["queryType"],
                query_result_type=data["queryResultType"],
                as_of=data["asOf"],
                columns=[
                    DevOpsWiqlColumn(
                        reference_name=column["referenceName"],
                        name=column["name"],
                        url=column["url"],
                    )
                    for column in data["columns"]
                ],
                work_items=[
                    DevOpsWiqlWorkItem(
                        id=work_item["id"],
                        url=work_item["url"],
                    )
                    for work_item in data["workItems"]
                ],
            )

    async def get_work_items(
        self,
        organization: str,
        project: str,
        ids: list[int],
    ) -> DevOpsWorkItem:
        """Get Azure DevOps work items."""
        async with aiohttp.ClientSession() as session:
            response: aiohttp.ClientResponse = await self.get(
                session,
                f"https://dev.azure.com/{organization}/{project}/_apis/wit/workitems?ids={','.join(str(id) for id in ids)}&api-version=6.0",
            )
            if response.status != 200:
                return None
            data = await response.json()
            if data is None:
                return None

            return DevOpsWorkItem(
                count=data["count"],
                value=[
                    DevOpsWorkItemValue(
                        id=work_item["id"],
                        rev=work_item["rev"],
                        fields=DevOpsWorkItemValueFields(
                            area_path=work_item["fields"]["System.AreaPath"],
                            team_project=work_item["fields"]["System.TeamProject"],
                            iteration_path=work_item["fields"]["System.IterationPath"],
                            work_item_type=work_item["fields"]["System.WorkItemType"],
                            state=work_item["fields"]["System.State"],
                            reason=work_item["fields"]["System.Reason"],
                            assigned_to=DevOpsWorkItemUser(
                                display_name=work_item["fields"]["System.AssignedTo"][
                                    "displayName"
                                ],
                                url=work_item["fields"]["System.AssignedTo"]["url"],
                                links=DevOpsWorkItemLinks(
                                    avatar=DevOpsWorkItemAvatar(
                                        href=work_item["fields"]["System.AssignedTo"]["_links"][
                                            "avatar"
                                        ]["href"],
                                    ),
                                ),
                                id=work_item["fields"]["System.AssignedTo"]["id"],
                                unique_name=work_item["fields"]["System.AssignedTo"]["uniqueName"],
                                image_url=work_item["fields"]["System.AssignedTo"]["imageUrl"],
                                descriptor=work_item["fields"]["System.AssignedTo"]["descriptor"],
                            )
                            if "System.AssignedTo" in work_item["fields"]
                            and work_item["fields"]["System.AssignedTo"]
                            else None,
                            created_date=work_item["fields"]["System.CreatedDate"],
                            created_by=DevOpsWorkItemUser(
                                display_name=work_item["fields"]["System.CreatedBy"]["displayName"],
                                url=work_item["fields"]["System.CreatedBy"]["url"],
                                links=DevOpsWorkItemLinks(
                                    avatar=DevOpsWorkItemAvatar(
                                        href=work_item["fields"]["System.CreatedBy"]["_links"][
                                            "avatar"
                                        ]["href"],
                                    ),
                                ),
                                id=work_item["fields"]["System.CreatedBy"]["id"],
                                unique_name=work_item["fields"]["System.CreatedBy"]["uniqueName"],
                                image_url=work_item["fields"]["System.CreatedBy"]["imageUrl"],
                                descriptor=work_item["fields"]["System.CreatedBy"]["descriptor"],
                            )
                            if "System.CreatedBy" in work_item["fields"]
                            and work_item["fields"]["System.CreatedBy"]
                            else None,
                            changed_date=work_item["fields"]["System.ChangedDate"],
                            changed_by=DevOpsWorkItemUser(
                                display_name=work_item["fields"]["System.ChangedBy"]["displayName"],
                                url=work_item["fields"]["System.ChangedBy"]["url"],
                                links=DevOpsWorkItemLinks(
                                    avatar=DevOpsWorkItemAvatar(
                                        href=work_item["fields"]["System.ChangedBy"]["_links"][
                                            "avatar"
                                        ]["href"],
                                    ),
                                ),
                                id=work_item["fields"]["System.ChangedBy"]["id"],
                                unique_name=work_item["fields"]["System.ChangedBy"]["uniqueName"],
                                image_url=work_item["fields"]["System.ChangedBy"]["imageUrl"],
                                descriptor=work_item["fields"]["System.ChangedBy"]["descriptor"],
                            )
                            if "System.ChangedBy" in work_item["fields"]
                            and work_item["fields"]["System.ChangedBy"]
                            else None,
                            comment_count=work_item["fields"]["System.CommentCount"],
                            title=work_item["fields"]["System.Title"],
                            microsoft_vsts_common_state_change_date=work_item["fields"],
                            microsoft_vsts_common_priority=work_item["fields"][
                                "Microsoft.VSTS.Common.Priority"
                            ],
                        ),
                        url=work_item["url"],
                    )
                    for work_item in data["value"]
                ],
            )
