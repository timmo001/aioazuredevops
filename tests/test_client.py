"""Test the http client module."""

# from aioresponses import aioresponses
from aioresponses import aioresponses
import pytest

from aioazuredevops.builds import DevOpsBuild
from aioazuredevops.client import BASE_URL, DevOpsClient
from aioazuredevops.core import DevOpsProject
from aioazuredevops.wiql import DevOpsWiqlResult
from aioazuredevops.work_item import (
    DevOpsWorkItem,
    DevOpsWorkItemValue,
    DevOpsWorkItemValueFields,
)

from . import ORGANIZATION, PAT, PROJECT

BAD_PROJECT_NAME = "badproject"
EMPTY_PROJECT_NAME = "emptyproject"


@pytest.mark.asyncio
async def test_authorize(
    mock_aioresponse: aioresponses,
    devops_client: DevOpsClient,
) -> None:
    """Test the authorize method."""
    assert not devops_client.authorized
    assert devops_client.pat is None

    assert await devops_client.authorize(
        pat=PAT,
        organization=ORGANIZATION,
    )

    assert devops_client.authorized
    assert devops_client.pat == PAT

    # Test with unauthorized (GET with PAT)
    mock_aioresponse.clear()
    mock_aioresponse.get(
        f"{BASE_URL}/{ORGANIZATION}/_apis/projects",
        status=401,
    )

    assert not await devops_client.authorize(
        pat=PAT,
        organization=ORGANIZATION,
    )

    assert not devops_client.authorized
    assert devops_client.pat == PAT


@pytest.mark.asyncio
async def test_get_project(
    mock_aioresponse: aioresponses,
    devops_client: DevOpsClient,
) -> None:
    """Test the get_project method."""
    project = await devops_client.get_project(
        organization=ORGANIZATION,
        project=PROJECT,
    )

    assert isinstance(project, DevOpsProject)
    assert project.project_id == "testid"

    # Test with authorization (GET with PAT)
    assert await devops_client.authorize(
        pat=PAT,
        organization=ORGANIZATION,
    )

    project = await devops_client.get_project(
        organization=ORGANIZATION,
        project=PROJECT,
    )

    # Test with client error
    mock_aioresponse.get(
        f"{BASE_URL}/{ORGANIZATION}/_apis/projects/{BAD_PROJECT_NAME}",
        status=400,
    )

    bad_project = await devops_client.get_project(
        organization=ORGANIZATION,
        project=BAD_PROJECT_NAME,
    )

    assert bad_project is None

    # Test with empty response
    mock_aioresponse.get(
        f"{BASE_URL}/{ORGANIZATION}/_apis/projects/{EMPTY_PROJECT_NAME}",
        payload=None,
        status=200,
    )

    empty_project = await devops_client.get_project(
        organization=ORGANIZATION,
        project=EMPTY_PROJECT_NAME,
    )

    assert empty_project is None


@pytest.mark.asyncio
async def test_get_builds(
    mock_aioresponse: aioresponses,
    devops_client: DevOpsClient,
) -> None:
    """Test the get_builds method."""
    builds = await devops_client.get_builds(
        organization=ORGANIZATION,
        project=PROJECT,
        parameters="",
    )

    assert isinstance(builds, list)
    assert len(builds) == 1
    assert isinstance(builds[0], DevOpsBuild)

    # Test with bad request
    mock_aioresponse.get(
        f"{BASE_URL}/{ORGANIZATION}/{BAD_PROJECT_NAME}/_apis/build/builds",
        status=400,
    )

    bad_builds = await devops_client.get_builds(
        organization=ORGANIZATION,
        project=BAD_PROJECT_NAME,
        parameters="",
    )

    assert bad_builds is None

    # Test with empty response
    mock_aioresponse.get(
        f"{BASE_URL}/{ORGANIZATION}/{EMPTY_PROJECT_NAME}/_apis/build/builds",
        payload=None,
        status=200,
    )

    empty_builds = await devops_client.get_builds(
        organization=ORGANIZATION,
        project=EMPTY_PROJECT_NAME,
        parameters="",
    )

    assert empty_builds is None


@pytest.mark.asyncio
async def test_get_build(
    mock_aioresponse: aioresponses,
    devops_client: DevOpsClient,
) -> None:
    """Test the get_build method."""
    build = await devops_client.get_build(
        organization=ORGANIZATION,
        project=PROJECT,
        build_id=1,
    )

    assert isinstance(build, DevOpsBuild)

    # Test with bad request
    mock_aioresponse.get(
        f"{BASE_URL}/{ORGANIZATION}/{BAD_PROJECT_NAME}/_apis/build/builds/1",
        status=400,
    )

    bad_build = await devops_client.get_build(
        organization=ORGANIZATION,
        project=BAD_PROJECT_NAME,
        build_id=1,
    )

    assert bad_build is None

    # Test with empty response
    mock_aioresponse.get(
        f"{BASE_URL}/{ORGANIZATION}/{EMPTY_PROJECT_NAME}/_apis/build/builds/1",
        payload=None,
        status=200,
    )

    empty_build = await devops_client.get_build(
        organization=ORGANIZATION,
        project=EMPTY_PROJECT_NAME,
        build_id=1,
    )

    assert empty_build is None


@pytest.mark.asyncio
async def test_get_work_items_ids_all(
    mock_aioresponse: aioresponses,
    devops_client: DevOpsClient,
) -> None:
    """Test the get_work_items_ids method."""
    work_items_ids = await devops_client.get_work_items_ids_all(
        organization=ORGANIZATION,
        project=PROJECT,
    )

    assert isinstance(work_items_ids, DevOpsWiqlResult)

    # Test with authorization (POST with PAT)
    assert await devops_client.authorize(
        pat=PAT,
        organization=ORGANIZATION,
    )

    work_items_ids = await devops_client.get_work_items_ids_all(
        organization=ORGANIZATION,
        project=PROJECT,
    )

    assert isinstance(work_items_ids, DevOpsWiqlResult)

    # Test with client error
    mock_aioresponse.post(
        f"{BASE_URL}/{ORGANIZATION}/{BAD_PROJECT_NAME}/_apis/wit/wiql?api-version=6.0",
        status=400,
    )

    bad_work_items_ids = await devops_client.get_work_items_ids_all(
        organization=ORGANIZATION,
        project=BAD_PROJECT_NAME,
    )

    assert bad_work_items_ids is None

    # Test with empty response
    mock_aioresponse.post(
        f"{BASE_URL}/{ORGANIZATION}/{EMPTY_PROJECT_NAME}/_apis/wit/wiql?api-version=6.0",
        payload=None,
        status=200,
    )

    empty_work_items_ids = await devops_client.get_work_items_ids_all(
        organization=ORGANIZATION,
        project=EMPTY_PROJECT_NAME,
    )

    assert empty_work_items_ids is None


@pytest.mark.asyncio
async def test_get_work_items(
    mock_aioresponse: aioresponses,
    devops_client: DevOpsClient,
) -> None:
    """Test the get_work_items method."""
    work_items = await devops_client.get_work_items(
        organization=ORGANIZATION,
        project=PROJECT,
        ids=[1],
    )

    assert isinstance(work_items, DevOpsWorkItem)
    assert work_items.count == 1
    assert len(work_items.value) == 1
    assert isinstance(work_items.value[0], DevOpsWorkItemValue)
    assert isinstance(work_items.value[0].fields, DevOpsWorkItemValueFields)

    # Test bad request
    mock_aioresponse.get(
        f"{BASE_URL}/{ORGANIZATION}/{BAD_PROJECT_NAME}/_apis/wit/workitems?ids=1&api-version=6.0",
        status=400,
    )

    bad_work_items = await devops_client.get_work_items(
        organization=ORGANIZATION,
        project=BAD_PROJECT_NAME,
        ids=[1],
    )

    assert bad_work_items is None

    # Test with empty response
    mock_aioresponse.get(
        f"{BASE_URL}/{ORGANIZATION}/{EMPTY_PROJECT_NAME}/_apis/wit/workitems?ids=1&api-version=6.0",
        payload=None,
        status=200,
    )

    empty_work_items = await devops_client.get_work_items(
        organization=ORGANIZATION,
        project=EMPTY_PROJECT_NAME,
        ids=[1],
    )

    assert empty_work_items is None
