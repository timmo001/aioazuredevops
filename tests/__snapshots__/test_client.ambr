# serializer version: 1
# name: test_get_build[build]
  Build(build_id=1, build_number='testbuildnumber', status='teststatus', result='testresult', source_branch='testsourcebranch', source_version='testsourceversion', priority='testpriority', reason='testreason', queue_time='testqueuetime', start_time='teststarttime', finish_time='teststarttime', definition=BuildDefinition(build_id=1, name='testname', url='testurl', path='testpath', build_type='testtype', queue_status='testqueuestatus', revision=1), project=Project(id='testid', name='testname', description='testdescription', url='testurl', state='teststate', revision=1, visibility='testvisibility', capabilities=None, links=None, default_team=None, last_update_time=None), links=BuildLinks(l_self='testself', web='testweb', source_version_display_uri='testsourceversiondisplayuri', timeline='testtimeline', badge='testbadge'))
# ---
# name: test_get_builds[builds]
  list([
    Build(build_id=1, build_number='testbuildnumber', status='teststatus', result='testresult', source_branch='testsourcebranch', source_version='testsourceversion', priority='testpriority', reason='testreason', queue_time='testqueuetime', start_time='teststarttime', finish_time='teststarttime', definition=BuildDefinition(build_id=1, name='testname', url='testurl', path='testpath', build_type='testtype', queue_status='testqueuestatus', revision=1), project=Project(id='testid', name='testname', description='testdescription', url='testurl', state='teststate', revision=1, visibility='testvisibility', capabilities=None, links=None, default_team=None, last_update_time=None), links=BuildLinks(l_self='testself', web='testweb', source_version_display_uri='testsourceversiondisplayuri', timeline='testtimeline', badge='testbadge')),
  ])
# ---
# name: test_get_project[project]
  Project(id='testid', name='testname', description='testdescription', url='testurl', state='teststate', revision=1, visibility='testvisibility', capabilities=Capabilities(process_template=ProcessTemplate(template_name='Agile', template_type_id='abc123-abc1-abc1-abc1-abc123456789'), versioncontrol=VersionControl(source_control_type='Git', git_enabled='True', tfvc_enabled='False')), links=Links(links_self=LinkCollection(href='testself'), collection=LinkCollection(href='testcollection'), web=LinkCollection(href='testweb')), default_team=DefaultTeam(id='testid', name='testname', url='testurl'), last_update_time=None)
# ---
# name: test_get_work_item_types[work_item_types]
  list([
    WorkItemType(name='testname', reference_name='testreferenceName', description='testdescription', color='testcolor', icon=Icon(id='testid', url='testurl'), is_disabled=False, xml_form='testxmlForm', fields=[], field_instances=[], transitions={}, states=[State(name='New', color='b2b2b2', category=<Category.PROPOSED: 'Proposed'>), State(name='On Hold', color='fcfea8', category=<Category.IN_PROGRESS: 'InProgress'>), State(name='Ready', color='007acc', category=<Category.IN_PROGRESS: 'InProgress'>), State(name='In Development', color='e87025', category=<Category.IN_PROGRESS: 'InProgress'>), State(name='In Code Review', color='f9b978', category=<Category.IN_PROGRESS: 'InProgress'>), State(name='In Test', color='fbd144', category=<Category.IN_PROGRESS: 'InProgress'>), State(name='In UAT', color='71338d', category=<Category.IN_PROGRESS: 'InProgress'>), State(name='In PreProd', color='ef33a3', category=<Category.IN_PROGRESS: 'InProgress'>), State(name='Ready for Release', color='207752', category=<Category.IN_PROGRESS: 'InProgress'>), State(name='Released', color='c3d84c', category=<Category.RESOLVED: 'Resolved'>), State(name='Closed', color='339933', category=<Category.COMPLETED: 'Completed'>)], url='testurl'),
  ])
# ---
# name: test_get_work_item_types[work_item_types_authorized]
  list([
    WorkItemType(name='testname', reference_name='testreferenceName', description='testdescription', color='testcolor', icon=Icon(id='testid', url='testurl'), is_disabled=False, xml_form='testxmlForm', fields=[], field_instances=[], transitions={}, states=[State(name='New', color='b2b2b2', category=<Category.PROPOSED: 'Proposed'>), State(name='On Hold', color='fcfea8', category=<Category.IN_PROGRESS: 'InProgress'>), State(name='Ready', color='007acc', category=<Category.IN_PROGRESS: 'InProgress'>), State(name='In Development', color='e87025', category=<Category.IN_PROGRESS: 'InProgress'>), State(name='In Code Review', color='f9b978', category=<Category.IN_PROGRESS: 'InProgress'>), State(name='In Test', color='fbd144', category=<Category.IN_PROGRESS: 'InProgress'>), State(name='In UAT', color='71338d', category=<Category.IN_PROGRESS: 'InProgress'>), State(name='In PreProd', color='ef33a3', category=<Category.IN_PROGRESS: 'InProgress'>), State(name='Ready for Release', color='207752', category=<Category.IN_PROGRESS: 'InProgress'>), State(name='Released', color='c3d84c', category=<Category.RESOLVED: 'Resolved'>), State(name='Closed', color='339933', category=<Category.COMPLETED: 'Completed'>)], url='testurl'),
  ])
# ---
# name: test_get_work_items[work_items]
  list([
    WorkItem(id=1, rev=234, fields=WorkItemFields(area_path='testAreaPath', team_project='testTeamProject', iteration_path='testIterationPath', work_item_type='testWorkItemType', state='testState', reason='testReason', assigned_to=None, created_date='testCreatedDate', created_by=None, changed_date='testChangedDate', changed_by=None, comment_count=1, title='testTitle', microsoft_vsts_common_state_change_date=None, microsoft_vsts_common_priority=1), url='testurl'),
    WorkItem(id=1, rev=234, fields=WorkItemFields(area_path='testAreaPath', team_project='testTeamProject', iteration_path='testIterationPath', work_item_type='testWorkItemType', state='testState', reason='testReason', assigned_to=None, created_date='testCreatedDate', created_by=None, changed_date='testChangedDate', changed_by=None, comment_count=1, title='testTitle', microsoft_vsts_common_state_change_date=None, microsoft_vsts_common_priority=1), url='testurl'),
  ])
# ---
# name: test_get_work_items_ids[work_items_ids]
  list([
    1,
  ])
# ---
# name: test_get_work_items_ids[work_items_ids_authorized]
  list([
    1,
  ])
# ---
